sdk/core/core-https/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -33,11 +42,39 @@ export function bearerTokenAuthenticationPolicy(
   options: BearerTokenAuthenticationPolicyOptions
 ): PipelinePolicy {
   const { credential, scopes } = options;
-  const tokenCache: AccessTokenCache = new ExpiringAccessTokenCache();
+  const tokenCache: ExpiringAccessTokenCache = new ExpiringAccessTokenCache(
+    options.tokenRefreshBufferMs
+  );
+
+  let refreshPromise: Promise<AccessToken | null> | undefined;
+
+  let refreshAttemptsBufferMs = options.refreshAttemptsBufferMs || 30000;
+
+  // When the refresh buffer is hit, we trigger this refresher
+  // which tries to update the token every so often.
+  // The delay of the refreshes is defined by the `refreshAttemptsBufferMs`
+  async function refresher(tokenOptions: GetTokenOptions): Promise<AccessToken | null> {
+    // If the cached token expired, we should stop refreshing the token.
+    while (!tokenCache.isExpired()) {
+      const token = await credential.getToken(scopes, tokenOptions);
+      if (token) {
+        return token;
```

Comment:
I'm very confused by this. Promises only have a single settled state, so wouldn't this `return` complete the promise? Then it would stop running and not try to refresh at all? Not sure how async functions handle while loops with a return in them.

Reply:
once it returns we've refreshed the token 🌞 so we should not continue to attempt getting tokens until the token is about to expire again 🤔 that's the idea I'm trying to portray.

Reply:
Oh derp. So you keep calling getToken() until it returns a truthy value? I see it now. Seems like there should maybe a max timeout though

Reply:
Right now in this code the max timeout is the token expiration date! Is that ok?